src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -286,12 +287,34 @@ unsafe long GetFileLengthCore()
             {
                 Interop.Kernel32.FILE_STANDARD_INFO info;
 
-                if (!Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
+                if (Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
+                {
+                    return info.EndOfFile;
+                }
+
+                if (_path is null || !PathInternal.IsDevice(_path))
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error(Path);
+                }
+
+                Interop.Kernel32.STORAGE_READ_CAPACITY storageReadCapacity;
+
+                bool success = Interop.Kernel32.DeviceIoControl(
+                    this,
+                    dwIoControlCode: Interop.Kernel32.IOCTL_STORAGE_READ_CAPACITY,
+                    lpInBuffer: null,
+                    nInBufferSize: 0,
+                    lpOutBuffer: &storageReadCapacity,
+                    nOutBufferSize: (uint)sizeof(Interop.Kernel32.STORAGE_READ_CAPACITY),
+                    out _,
```

Comment:
Do we need to validate that we have read `sizeof(Interop.Kernel32.STORAGE_READ_CAPACITY)` bytes to avoid reading garbage below?

Reply:
> Do we need to validate that we have read sizeof(Interop.Kernel32.STORAGE_READ_CAPACITY) bytes to avoid reading garbage below?

@jkotas sure! But we still need to throw. If sys-call `GetFileInformationByHandleEx` falls with an error and then we call `DeviceIoControl` which does not fail, but returns fewer bytes than requested is the last error still going to be set with an error from the first sys-call?

Reply:
The last error is set to SUCCESS ;) Give me a second